gridnet/connections/connections.py

```diff
@@ -0,0 +1,137 @@
+import tensorflow as tf
+from common.models import ConvNetwork
+from utils.misc import tf_coin_flip, print_tensor_shape
+
+class LateralConnection(ConvNetwork):
+    def __init__(self, name, layer_specs,
+                 activation_fn=tf.nn.leaky_relu,
+                 total_dropout_rate=0.0,
+                 regularizer=None):
+        """
+        :param name: Str. For variable scoping.
+        :param layer_specs: Array of shape [num_layers, 2]. Constrained version of parent class' layer_specs.
+                            The second dimension consists of [num_output_features, dilation].
+        :param activation_fn: Tensorflow activation function. This will not be applied on the last convolutional layer.
+        :param total_dropout_rate: A value of 1.0 will always zero-out the output of this block, and 0.0 will keep it.
+        :param regularizer: Tf regularizer such as tf.contrib.layers.l2_regularizer.
+        """
+        full_layer_specs = []
+        for i, layer_spec in enumerate(layer_specs):
+            full_layer_spec = [3, layer_spec[0], layer_spec[1], 1]
+            full_layer_specs.append(full_layer_spec)
+
+        super().__init__(layer_specs=full_layer_specs,
+                         activation_fn=activation_fn, last_activation_fn=None,
+                         regularizer=regularizer, padding='SAME')
+
+        self.name = name
+        self.total_dropout_rate = total_dropout_rate
+
+    def get_forward(self, features, reuse_variables=False, training=False):
+        """
+        :param features: Tensor. Feature map of shape [batch_size, H, W, num_features].
+        :param reuse_variables: Bool. Whether to reuse the variables.
+        :return: Tensor. Feature map of shape [batch_size, H, W, num_features].
+        """
+        with tf.variable_scope(self.name, reuse=reuse_variables):
+
+            # Pass through resolution preserving convolutions, with skip connection at the end.
+            if self.activation_fn is not None:
+                previous_output = self.activation_fn(features)
+
+            previous_output, layer_outputs = self._get_conv_tower(previous_output)
+            final_output = features + previous_output
```

Comment:
Is this supposed to be an addition or a concatenation? I can't tell from the context-aware paper.

Reply:
These use the same ops as the grid intersections outputs -- they are additions.